workflow/controller/dag.go

```diff
@@ -632,6 +629,17 @@ func expandTask(task wfv1.DAGTask) ([]wfv1.DAGTask, error) {
 		return []wfv1.DAGTask{task}, nil
 	}
 
+	taskBytes, err := json.Marshal(task)
+	if err != nil {
+		return nil, errors.InternalWrapError(err)
+	}
+
+	// these fields can be very large (>100m) and marshalling 10k x 100m = 6GB of memory used and
+	// very poor performance, so we just nil them out
+	task.WithItems = nil
+	task.WithParam = ""
+	task.WithSequence = nil
```

Comment:
`task` is not a pointer, but a value and these three fields are not referenced elsewhere in this function. Setting these values to `nil` seems inconsequential?

Reply:
Not at all. The problem, which maybe I don't clearly state, is that these redundant fields are very large if you have many params/items. This makes JSON marshalling very memory intensive at O(n^2). By nilling them out, it save this memory.

Reply:
This heap graph shows the problem memory usage dominated by JSON marshalling:
![](https://user-images.githubusercontent.com/1142830/94880401-ad1c3d00-0417-11eb-93f1-e675618512fd.png)

After the change:


![](https://user-images.githubusercontent.com/1142830/94886478-21f77300-0428-11eb-839e-cd44a886c2c6.png)

Reply:
I understand the problem. What doesn't make sense to me is that the niling only happens in the local scope of this `expandTask` function, and since there is no Marshaling done (or in fact anything that considers these fields) in the rest of the function it seems to me that the niling has no effect.

Have you run a before/after comparison of the heap graph when adding this code?

Reply:
yes - see above - but there is not harm in double-checking - the marshalling is not done in this function, it in done when this func invokes processItem (operator.go#2699).

You can actually see this is the first graph.